stan.go

```diff
@@ -0,0 +1,646 @@
+	// Furthermore, the decimal separator can occur only in positions 12, 20, 28, corresponding to
+	// whether temperature is negative, and contains one (or two) decimals before the separator.
+	// Thus, the following mask suffices to clear all decimals and keep the decimal separator bit.
+	return bits.TrailingZeros64(uint64(^word & 0x10101000))
+}
+
+//gcassert:inline
+func convertIntoNumber(decimalSepPos int, word int64) int16 {
+	shift := 28 - decimalSepPos
+	// signed is -1 if negative, 0 otherwise
+	signed := (^word << 59) >> 63
+	designMask := ^(signed & 0xFF)
+	// Align the number to a specific position and transform the ascii code
+	// to actual digit value in each byte
+	digits := ((word & designMask) << shift) & int64(0x0F000F0F00)
+
+	// Now digits is in the form 0xUU00TTHH00 (UU: units digit, TT: tens digit, HH: hundreds digit)
+	// 0xUU00TTHH00 * (100 * 0x1000000 + 10 * 0x10000 + 1) =
+	// 0x000000UU00TTHH00 +
+	// 0x00UU00TTHH000000 * 10 +
+	// 0xUU00TTHH00000000 * 100
+	// Now TT * 100 has 2 trailing zeroes and HH * 100 + TT * 10 + UU < 0x400
+	// This results in our value lies in the bit 32 to 41 of this product
+	// That was close :)
+	absValue := ((digits * 0x640a0001) >> 32) & 0x3FF
+	value := (absValue ^ signed) - signed
+	return int16(value)
+}
+
+//gcassert:inline
+func maskWord(word int64, semiclnWordBytePos int) int64 {
+	// Shift 1s into [0, semiclnWordBytePos] to form the mask.
+	return word & int64(^uint(0)>>(semiclnWordBytePos<<3))
+}
+
+// Returns the index of the hash table entry for the given (key) hash.
+// requires: sizePowerOfTwo is a power of two
+//
+//gcassert:inline
+func hashToIndex(hash int64, sizePowerOfTwo int) int {
+	hash32 := int(hash ^ (hash >> 28))
+	finalHash := hash32 ^ (hash32 >> 15)
+	// N.B. sizePowerOfTwo is a power of two, so the following is equivalent to finalHash % sizePowerOfTwo.
+	return finalHash & (sizePowerOfTwo - 1)
+}
+
+type PartitionResult struct {
+	nameLength                  int
+	min, max                    int16
+	firstNamePart, lastNamePart int64
+	namePos                     int64
+	count, sum                  int64
+}
+
+type AggResult struct {
+	min, max   int16
+	count, sum int64
+	name       string
+}
+
+//gcassert:inline
+func merge(a *AggResult, b *AggResult) {
+	b.count += a.count
+	b.sum += a.sum
+	b.min = min(a.min, b.min)
+	b.max = max(a.max, b.max)
+}
+
+//gcassert:inline
+func newEntry(inputRowPos int64, stationNameLen int, temperature int16, firstNamePart, lastNamePart int64) *PartitionResult {
+	entry := &PartitionResult{
+		namePos:       inputRowPos,
+		nameLength:    stationNameLen,
+		min:           temperature,
+		max:           temperature,
+		count:         1,
+		sum:           int64(temperature),
+		firstNamePart: firstNamePart,
+		lastNamePart:  lastNamePart,
+	}
+	return entry
+}
+
+//gcassert:inline
+func updateEntry(entry *PartitionResult, number int16) {
+	entry.sum += int64(number)
+	if number < entry.min {
+		entry.min = number
+		return
```

Comment:
bug? The `entry.count++` is skipped in this case.

Reply:
🤦 yeah, I literally changed the counter increment order in the last hour, after swar optimization didn't pan out.